source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+  }
+  post_connect_state_.read_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->removeReadFilter(filter);
+    return;
+  }
+  auto i = post_connect_state_.read_filters_.begin();
+  while (i != post_connect_state_.read_filters_.end()) {
+    if (*i == filter) {
+      post_connect_state_.read_filters_.erase(i);
+      return;
+    }
+  }
+  ASSERT(false);
+}
+
+bool HappyEyeballsConnectionImpl::initializeReadFilters() {
+  if (connect_finished_) {
+    return connections_[0]->initializeReadFilters();
+  }
+  if (!post_connect_state_.read_filters_.empty()) {
+    return false;
+  }
+  post_connect_state_.initialize_read_filters_.value() = true;
+  return true;
+}
+
+void HappyEyeballsConnectionImpl::addBytesSentCallback(Connection::BytesSentCb cb) {
+  if (connect_finished_) {
+    connections_[0]->addBytesSentCallback(cb);
+    return;
+  }
+  post_connect_state_.bytes_sent_callbacks_.push_back(cb);
+}
+
+void HappyEyeballsConnectionImpl::enableHalfClose(bool enabled) {
+  if (!connect_finished_) {
+    per_connection_state_.enable_half_close_ = enabled;
+  }
+  for (auto& connection : connections_) {
+    connection->enableHalfClose(enabled);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::isHalfCloseEnabled() {
+  if (connect_finished_) {
+    return connections_[0]->isHalfCloseEnabled();
+  }
+  return per_connection_state_.enable_half_close_.has_value() &&
+         per_connection_state_.enable_half_close_.value();
+}
+
+std::string HappyEyeballsConnectionImpl::nextProtocol() const {
+  if (connect_finished_) {
+    return connections_[0]->nextProtocol();
+  }
+  return "";
+}
+
+void HappyEyeballsConnectionImpl::noDelay(bool enable) {
+  if (!connect_finished_) {
+    per_connection_state_.no_delay_ = enable;
+  }
+  for (auto& connection : connections_) {
+    connection->noDelay(enable);
+  }
+}
+
+void HappyEyeballsConnectionImpl::readDisable(bool disable) {
+  if (connect_finished_) {
+    connections_[0]->readDisable(disable);
+    return;
+  }
+  if (!post_connect_state_.read_disable_count_.has_value()) {
+    post_connect_state_.read_disable_count_ = 0;
+  }
+
+  if (disable) {
+    post_connect_state_.read_disable_count_.value()++;
+  } else {
+    ASSERT(post_connect_state_.read_disable_count_ != 0);
+    post_connect_state_.read_disable_count_.value()--;
+  }
+}
+
+void HappyEyeballsConnectionImpl::detectEarlyCloseWhenReadDisabled(bool value) {
+  if (!connect_finished_) {
+    per_connection_state_.detect_early_close_when_read_disabled_ = value;
+  }
+  for (auto& connection : connections_) {
+    connection->detectEarlyCloseWhenReadDisabled(value);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::readEnabled() const {
+  if (!connect_finished_) {
+    return !post_connect_state_.read_disable_count_.has_value() ||
+           post_connect_state_.read_disable_count_ == 0;
+  }
+  return connections_[0]->readEnabled();
```

Comment:
also if we pass on immediately we can always return the value of connections_[0]